students/jjakub/session07/html_render/test_html_render.py

```diff
@@ -0,0 +1,334 @@
+    #       properly rendered, so we don't need to test that here.
+    assert "some plain text" in file_contents
+    assert "A simple paragraph of text" in file_contents
+    assert "Some more plain text." in file_contents
+    assert "some plain text" in file_contents
+    # but make sure the embedded element's tags get rendered!
+    assert "<p>" in file_contents
+    assert "</p>" in file_contents
+    # assert False
+
+
+
+########
+# Step 3
+########
+
+def test_head():
+    """
+    tests that you can render a single line header 
+    """
+    e = Head("this is some text")
+    e.append("and this is some more text")
+
+    file_contents = render_result(e).strip()
+
+    assert("this is some text") in file_contents
+    assert("and this is some more text") in file_contents
+
+    assert file_contents.startswith("<head>")
+    assert file_contents.endswith("</head>")
+
+
+def test_title():
+    t = Title()
+    t.append("This is the title")
+
+    file_contents = render_result(t)
+    print(file_contents)
+    assert("This is the title") in file_contents
+    assert file_contents.startswith("<title>")
+    # assert file_contents.endswith("</title>")
+    # assert "\n" not in file_contents
+
+
+def test_attributes():
+    e = P("A paragraph of text", style="text-align: center", id="intro")
+
+    file_contents = render_result(e).strip()
+    print(file_contents)  # so we can see it if the test fails
+
+    # note: The previous tests should make sure that the tags are getting
+    #       properly rendered, so we don't need to test that here.
+    #       so using only a "P" tag is fine
+    assert "A paragraph of text" in file_contents
+    # but make sure the embedded element's tags get rendered!
+    # first test the end tag is there -- same as always:
+    assert file_contents.endswith("</p>")
+
+    # but now the opening tag is far more complex
+    # but it starts the same:
+    assert file_contents.startswith("<p")
+    assert 'style="text-align: center"' in file_contents
+    assert 'id="intro"' in file_contents
+
+
+def test_hr():
+    """a simple horizontal rule with no attributes"""
+    hr = Hr()
+    file_contents = render_result(hr)
+    print(file_contents)
+    assert file_contents == '<hr />\n'
+
+
+def test_hr_attr():
+    """a horizontal rule with an attribute"""
+    hr = Hr(width=400)
+    file_contents = render_result(hr)
+    print(file_contents)
+    assert file_contents == '<hr width="400" />\n'
+
+
+#####################
+# indentation testing
+#  Uncomment for Step 9 -- adding indentation
+#####################
+
+
+def test_indent():
+    """
+    Tests that the indentation gets passed through to the renderer
+    """
+    html = Html("some content")
+    file_contents = render_result(html, ind="   ").rstrip()  #remove the end newline
+
+    # print(file_contents)
+    lines = file_contents.split("\n")
+    print(file_contents)
+    # assert lines[0].startswith("   <")
```

Comment:
I suppose this is where your other test failure occurs? Could you paste the output of your `print(file_contents)` statement verbatim here?

Reply:
<!DOCTYPE html>
<html>
    some content
</html>